### DISCUSSION TITLE
Support for SwiftUI

### DISCUSSION THREAD

**Discussion by vmanot** [rajdeep/proton#83] [Ideas]
Proton currently doesn't offer any SwiftUI ready APIs.

**Comment by vmanot** [comment]
@rajdeep I'm willing to work on this, let me know if this is within the scope of the project or not.

**Comment by rajdeep** [comment]
that'd be great @vmanot. I have not been able to manage any time lately. While I do hope to be able to spend more time in near future, I don't think I will be able to get to this any time soon. You may look into this and we can have a discussion once you have an idea on how you plan to approach this.

**Comment by vmanot** [comment]
@rajdeep I understand - I think I'll develop it as a separate package then (I don't want to diverge too far from `master` and at the same time intend to use it across a few projects). If/when it reaches maturity, you could consider absorbing it back into the main framework if appropriate.

**Comment by rajdeep** [comment]
sounds like a plan! thanks again!

**Comment by papajyp** [comment]
I have been working on a basic swiftui Rich Text Editor. I am lacking lists. I might be able to help. I am interested at least incorporating what you have into my app.

**Comment by rajdeep** [comment]
that's cool, @papajyp. Please feel free to share any ideas, suggestions or code that you think would help get SwiftUI support.

**Comment by writing-shed** [comment]
A good starting point for Proton would be a SwiftUI version of EditorView. At the moment I am hosting EditorView in UIViewRepresentable, which works, but is messy and precludes giving EditorView more SwiftUI features.

**Reply by vmanot** [reply]
@writing-shed I am actively working on this. I'm building a SwiftUI wrapper over Proton (using SwiftUIX), and wouldn't mind open sourcing it if others are interested.

**Reply by santi-g-s** [reply]
@vmanot That would be great. Would this make proton compatible with native macOS apps? i.e not using catalyst

**Reply by jabethfla** [reply]
@rajdeep yes! please open source it

**Comment by writing-shed** [comment]
I have already embedded Proton in a SwiftUI app that I’m developing and it works OK on IOS and MacOS. All I did was use UIViewRepresentable to create the EditorView and let the Coordinator class act as EditorViewDelegate. In all it is 170 lines of code. I have no doubt that Proton can be fully SwiftUIified, and I look forward to what Vatsal does with it.

**Reply by santi-g-s** [reply]
Does this support non-catalyst macOS apps? Wouldn't it have to be through NSViewRepresentable?

**Comment by writing-shed** [comment]
I just have a single app targeted to run on iphone, ipad and mac (via catalyst). I haven't tried a non catalyst mac os app. I doubt if it would compile. It only needs uiviewrepresentable if using SwiftUI 

Best
Keith

**Comment by santi-g-s** [comment]
Hi @vmanot! Any update on how this is going?

**Comment by niklasgrewe** [comment]
Hi @vmanot any news about proton and swiftui?

**Comment by bryan1anderson** [comment]
@rajdeep I recommend as a first step we work on a simple example that wraps EditorView in a UIViewRepresentable.
That's pretty straightforward.

I think showing how to do that in the Read me would go along way towards helping people get started.

Here's a sample of mine:
```
struct ProtonView: View {
    
    @Binding var attributedText: NSAttributedString
    @State var height: CGFloat = 0
    var body: some View {
        ProtonWrapperView(attributedText: $attributedText) { view in
            let height = view.systemLayoutSizeFitting(UIView.layoutFittingExpandedSize).height

            self.height = height
        }
        .frame(height: height)
    }
}

struct ProtonWrapperView: UIViewRepresentable {
    
    @Binding var attributedText: NSAttributedString
    let textDidChange: (EditorView) -> Void

    
    func makeUIView(context: Context) -> EditorView {
        let view = EditorView()
        view.becomeFirstResponder()
        view.attributedText = attributedText
        view.isScrollEnabled = false
        view.setContentCompressionResistancePriority(.required, for: .vertical)

        view.heightAnchor.constraint(greaterThanOrEqualToConstant: 300).isActive = true
        
        DispatchQueue.main.async {
            self.textDidChange(view)
        }
        
        EditorViewContext.shared.delegate = context.coordinator

        return view
    }
    
    func updateUIView(_ view: EditorView, context: Context) {

    }
    
    func makeCoordinator() -> Coordinator {
        Coordinator(self)
    }
    
    class Coordinator: NSObject, EditorViewDelegate {
        var parent: ProtonWrapperView
        
        init(_ parent: ProtonWrapperView) {
            self.parent = parent
        }
        
        func editor(_ editor: EditorView, didChangeTextAt range: NSRange) {
            editor.isScrollEnabled = false
            parent.attributedText = editor.attributedText
            DispatchQueue.main.async {
                self.parent.textDidChange(editor)
            }
        }
    }
}
```

**Reply by wtpalexander** [reply]
Agreed. Adding some SwiftUI examples similar to this would be helpful imo!
